File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/buffs/Shadows.java

package com.watabou.pixeldungeon.actors.buffs;

import com.nyrds.Packable;
import com.nyrds.platform.audio.Sample;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.pixeldungeon.ui.BuffIndicator;

import org.jetbrains.annotations.NotNull;

public class Shadows extends Invisibility {

	@Packable
	protected float left;

	@Override
	public boolean attachTo(@NotNull Char target ) {
		if (super.attachTo( target )) {
			Sample.INSTANCE.play( Assets.SND_MELD );
			Dungeon.observe();
			return true;
		} else {
			return false;
		}
	}
	
	@Override
	public void detach() {
		super.detach();
		Dungeon.observe();
	}
	
	@Override
	public boolean act() {
		spend( TICK * 2 );

		if (target.isAlive()) {
			if (--left <= 0 || target.visibleEnemies() > 0) {
				detach();
			}
		} else {
			detach();
		}
		
		return true;
	}
	
	public void prolong() {
		left = 2;
	}
	
	@Override
	public int icon() {
		return BuffIndicator.SHADOWS;
	}

	@Override
	public CharSprite.State charSpriteStatus() {
		return CharSprite.State.INVISIBLE;
	}
}

File: scripts/buffs/Shadows.lua
---
--- Shadows: short invisibility renewed by Foliage (RPD.CharUtils:refreshShadows),
--- broken by visible enemies (was actors/buffs/Shadows.java)
---
local RPD  = require "scripts/lib/commonClasses"
local buff = require "scripts/lib/buff"

return buff.init{
    desc  = function ()
        return {
            icon          = 13, -- BuffIndicator.SHADOWS
            name          = "ShadowsBuff_Name",
            info          = "ShadowsBuff_Info",
        }
    end,

    attachTo = function(self, buff, target)
        target:adjustInvisibility(1)
        return true
    end,

    attached = function(self, buff)
        RPD.playSound("snd_meld")
        buff.target:observe()
    end,

    detach = function(self, buff)
        buff.target:adjustInvisibility(-1)
        buff.target:observe()
    end,

    act = function(self, buff)
        buff:spend(2) -- Actor.TICK * 2
        local target = buff.target
        if target:isAlive() then
            self.data.left = (self.data.left or 0) - 1
            if self.data.left <= 0 or target:visibleEnemies() > 0 then
                buff:detach()
            end
        else
            buff:detach()
        end
    end,

    -- java entry: CharUtils.refreshShadows
    prolong = function(self, buff)
        self.data.left = 2
    end,

    charSpriteStatus = function(self, buff)
        return "INVISIBLE"
    end
}
